File: src/agents/orchestrator.js
/**
 * New Orchestrator
 * Simplified orchestration using 3 core agents:
 * - Router Agent: Clean query + determine intent
 * - Conversation Agent: RAG-powered Q&A
 * - Action Agent: Handle bookings/complaints/feedback
 */

const { StateGraph, END } = require('@langchain/langgraph');
const { routeQuery } = require('./router-agent');
const { handleInformationQuery } = require('./conversation-agent');
const { handleAction } = require('./action-agent');
const { getCompanyByPhone, generateCompanyContext } = require('../config/company-config');
const { query } = require('../lib/db');

// Main orchestrator state
class OrchestratorState {
  constructor() {
    this.originalQuery = '';
    this.refinedQuery = '';
    this.intent = '';
    this.actionType = null;
    this.response = '';
    this.companyId = '';
    this.companyContext = null;
    this.customerPhone = '';
    this.conversationHistory = [];
    this.actionData = {};
    this.isActionComplete = false;
    this.callId = null;
    this.errors = [];
  }
}

/**
 * Process a customer query through the agent system
 * @param {string} query - Customer's spoken query
 * @param {string} twilioNumber - The Twilio number that received the call
 * @param {string} customerPhone - Customer's phone number
 * @param {Array} conversationHistory - Previous messages in this call
 * @param {Object} actionData - Previously collected action data (for multi-turn)
 */
async function processQuery(
  queryText,
  twilioNumber,
  customerPhone,
  conversationHistory = [],
  actionData = {}
) {
  const startTime = Date.now();

  try {
    // Step 1: Get company config from Twilio number
    const company = await getCompanyByPhone(twilioNumber);

    if (!company) {
      console.warn(`No company found for Twilio number: ${twilioNumber}`);
      return {
        response: "I apologize, but this number is not configured. Please try again later.",
        intent: 'error',
        errors: ['Company not found for phone number']
      };
    }

    const companyContext = generateCompanyContext(company);

    // Step 2: Route the query
    const routeResult = await routeQuery(queryText, conversationHistory);

    const { refinedQuery, intent, actionType } = routeResult;

    console.log(`[Orchestrator] Intent: ${intent}, Action: ${actionType || 'N/A'}`);

    let response = '';
    let isActionComplete = false;
    let updatedActionData = actionData;
    let interactionId = null;

    // Step 3: Handle based on intent
    if (intent === 'information' || intent === 'unclear') {
      // Use conversation agent for information queries
      const conversationResult = await handleInformationQuery(
        refinedQuery,
        company.id,
        companyContext,
        conversationHistory
      );
      response = conversationResult.response;

    } else if (intent === 'action' && actionType) {
      // Use action agent for bookings/complaints/feedback
      const actionResult = await handleAction(
        refinedQuery,
        actionType,
        company.id,
        companyContext,
        customerPhone,
        conversationHistory,
        actionData
      );

      response = actionResult.response;
      updatedActionData = actionResult.collectedData;
      isActionComplete = actionResult.isComplete;
      interactionId = actionResult.interactionId;

    } else {
      // Fallback
      response = companyContext.fallbackMessage ||
        "I'm not sure I understood that. Could you please rephrase?";
    }

    const duration = Date.now() - startTime;
    console.log(`[Orchestrator] Processed in ${duration}ms`);

    return {
      response,
      refinedQuery,
      intent,
      actionType,
      actionData: updatedActionData,
      isActionComplete,
      interactionId,
      companyId: company.id,
      companyName: company.name,
      errors: routeResult.errors || []
    };

  } catch (error) {
    console.error('Orchestrator error:', error);
    return {
      response: "I apologize, but I'm experiencing technical difficulties. Please try again.",
      intent: 'error',
      errors: [`Orchestrator error: ${error.message}`]
    };
  }
}

/**
 * Create the orchestrator workflow for LangGraph visualization
 */
function createOrchestratorWorkflow() {
  const workflow = new StateGraph({
    channels: {
      originalQuery: {},
      refinedQuery: {},
      intent: {},
      actionType: {},
      response: {},
      companyId: {},
      companyContext: {},
      customerPhone: {},
      conversationHistory: {},
      actionData: {},
      isActionComplete: {},
      callId: {},
      errors: {}
    }
  });

  // Node: Route query
  workflow.addNode('router', async (state) => {
    const result = await routeQuery(state.originalQuery, state.conversationHistory);
    return {
      ...state,
      refinedQuery: result.refinedQuery,
      intent: result.intent,
      actionType: result.actionType,
      errors: [...state.errors, ...result.errors]
    };
  });

  // Node: Handle information queries
  workflow.addNode('conversation', async (state) => {
    const result = await handleInformationQuery(
      state.refinedQuery,
      state.companyId,
      state.companyContext,
      state.conversationHistory
    );
    return {
      ...state,
      response: result.response,
      errors: [...state.errors, ...result.errors]
    };
  });

  // Node: Handle action queries
  workflow.addNode('action', async (state) => {
    const result = await handleAction(
      state.refinedQuery,
      state.actionType,
      state.companyId,
      state.companyContext,
      state.customerPhone,
      state.conversationHistory,
      state.actionData
    );
    return {
      ...state,
      response: result.response,
      actionData: result.collectedData,
      isActionComplete: result.isComplete,
      errors: [...state.errors, ...result.errors]
    };
  });

  // Conditional routing from router
  workflow.addConditionalEdges(
    'router',
    (state) => {
      if (state.intent === 'action' && state.actionType) {
        return 'action';
      }
      return 'conversation';
    }
  );

  // Both conversation and action go to END
  workflow.addEdge('conversation', END);
  workflow.addEdge('action', END);

  // Entry point
  workflow.setEntryPoint('router');

  return workflow.compile();
}

/**
 * Get the initial greeting for a company
 */
async function getGreeting(twilioNumber) {
  const company = await getCompanyByPhone(twilioNumber);

  if (!company) {
    return "Hello, thank you for calling. How may I assist you today?";
  }

  return company.greeting || `Hello, thank you for calling ${company.name}. How may I assist you today?`;
}

module.exports = {
  processQuery,
  createOrchestratorWorkflow,
  getGreeting,
  OrchestratorState
};

File: backend/app/agents/orchestrator.py
"""
Orchestrator – Routes queries through the 3-agent pipeline.
"""

import logging
import time
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select

from app.agents.router_agent import route_query
from app.agents.conversation_agent import handle_information_query
from app.agents.action_agent import handle_action
from app.models.models import Company, CompanyService, Interaction
from app.utils.company_config import get_company_context

logger = logging.getLogger(__name__)


async def process_query(
    query_text: str,
    twilio_number: str,
    customer_phone: str,
    db: AsyncSession,
    conversation_history: list[dict] | None = None,
    action_data: dict | None = None,
) -> dict:
    """
    Main entry point – processes a customer query end-to-end.

    1. Look up company by Twilio number
    2. Route query (clean + classify)
    3. Handle via conversation or action agent
    """
    start = time.time()
    conversation_history = conversation_history or []
    action_data = action_data or {}

    # 1. Resolve company
    result = await db.execute(
        select(Company).where(
            Company.phone_number == twilio_number,
            Company.is_active.is_(True),
        )
    )
    company = result.scalar_one_or_none()

    if company is None:
        logger.warning("No company for Twilio number %s", twilio_number)
        return {
            "response": "I apologize, but this number is not configured. Please try again later.",
            "intent": "error",
            "errors": ["Company not found for phone number"],
        }

    ctx = await get_company_context(str(company.id), db)

    # 2. Route
    route_result = await route_query(query_text, conversation_history)
    intent = route_result["intent"]
    action_type = route_result.get("actionType")
    refined = route_result["refinedQuery"]

    logger.info("[Orchestrator] intent=%s action=%s", intent, action_type or "N/A")

    response = ""
    is_action_complete = False
    updated_action_data = action_data
    interaction_id = None

    # 3. Dispatch
    if intent in ("information", "unclear"):
        conv_result = await handle_information_query(
            refined, str(company.id), ctx, conversation_history
        )
        response = conv_result["response"]

    elif intent == "action" and action_type:
        act_result = await handle_action(
            refined,
            action_type,
            str(company.id),
            ctx,
            customer_phone,
            conversation_history,
            action_data,
        )
        response = act_result["response"]
        updated_action_data = act_result["collectedData"]
        is_action_complete = act_result["isComplete"]

        # Persist completed interaction
        if is_action_complete:
            interaction = Interaction(
                company_id=company.id,
                type=action_type,
                customer_name=updated_action_data.get("customerName", "Unknown"),
                customer_phone=customer_phone,
                data=updated_action_data,
                summary=_summary(action_type, updated_action_data),
            )
            db.add(interaction)
            await db.flush()
            interaction_id = str(interaction.id)
    else:
        response = ctx.get(
            "fallbackMessage",
            "I'm not sure I understood that. Could you please rephrase?",
        )

    elapsed = round((time.time() - start) * 1000)
    logger.info("[Orchestrator] processed in %dms", elapsed)

    return {
        "response": response,
        "refinedQuery": refined,
        "intent": intent,
        "actionType": action_type,
        "actionData": updated_action_data,
        "isActionComplete": is_action_complete,
        "interactionId": interaction_id,
        "companyId": str(company.id),
        "companyName": company.name,
        "errors": route_result.get("errors", []),
    }


async def get_greeting(twilio_number: str, db: AsyncSession) -> str:
    """Return the company greeting for a phone number."""
    result = await db.execute(
        select(Company).where(Company.phone_number == twilio_number)
    )
    company = result.scalar_one_or_none()
    if company is None:
        return "Hello, thank you for calling. How may I assist you today?"
    return company.greeting or f"Hello, thank you for calling {company.name}. How may I assist you today?"


def _summary(action_type: str, data: dict) -> str:
    if action_type == "booking":
        return f"Booking for {data.get('serviceId', '?')} on {data.get('preferredDate', '?')} at {data.get('preferredTime', '?')}"
    if action_type == "complaint":
        desc = data.get("issueDescription", "")
        return f"Issue: {desc[:100]}"
    if action_type == "feedback":
        fb = data.get("feedbackContent", "")
        rating = data.get("rating", "")
        return f"Feedback{f' ({rating}/5)' if rating else ''}: {fb[:100]}"
    return f"{action_type} from {data.get('customerName', 'Unknown')}"
